### PR TITLE
Issue #3241260: Event Date pushes back one day when someone edits the all-day event

### PR DESCRIPTION
**Pull request by rollki**
## Problem
When someone is editing the `all-day` event, that event can go back one day in time each time it is edited.

E.G. As a user with `Amsterdam time` timezone, I created an event, and then as a user with `New York` timezone I'll edit that event, the event date will be converted to `New York` timezone and that is correct, but after saving the date in the database will be changes to a new one and it's wrong because event goes back one day in time.

## Solution
The data that event is all day is not saved anywhere. So, here's two ways:

- We can add a custom field to the event (`field_event_all_day`)
- We can save that data to State API

Regarding the conversation belove, we decided to use a separate field in the events to set all day (first option).

## Issue tracker
- https://www.drupal.org/project/social/issues/3241260
- https://getopensocial.atlassian.net/browse/YANG-6332

_Related:_
- https://getopensocial.atlassian.net/browse/SUP-1544

## How to test
- [ ] Create the all-day event as a user with `Amsterdam time` timezone
- [ ] Go edit that event as a user with `New York` timezone
- [ ] Each time when you will edit that event under a user with New York timezone, that event will go back one day in time

See screenshots.

## Screenshots
image

Screenshot 2021-10-07 at 11 58 14

Screenshot 2021-10-07 at 11 58 43

## Release notes
No.

## Change Record
The method `social_event_date_is_all_day` was changed. New `social_event_date_set_all_day` method was added. 

## Translations
No.

### CODE REVIEW

**Comment by rollki** [comment]
These changes on the testing stage at the moment.

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2539-vab3fds8v2pqe6bjocasq64xdhgz5fpe.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/615eb3cb7207f224b434de84

**Comment by rollki** [comment]
**Note:** In Slack (https://opensocial.slack.com/archives/C358ZEJMS/p1633597587003200) we decided to reserve some extra time to resolve this issue in proper way (using the field API).

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2539-vab3fds8v2pqe6bjocasq64xdhgz5fpe.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/615eb3cb7207f224b434de84

**Comment by ribel** [comment]
@Kingdutch, @navneet0693, @Ressinel 
Looks like work on this PR is completed and it's ready to merge. Also, it's already used in production sites.
So do you have any objections against merging it?

**Comment by Ressinel** [comment]
@ribel Check my comment above. It is the only thing I mentioned.

**Comment by ribel** [comment]
@rollki I think it's a good point. Can you do that, so we can finally merge this PR?

**Comment by rollki** [comment]
@Ressinel @ribel 
Done, let's wait to tests green and merge I guess :)

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2539-vab3fds8v2pqe6bjocasq64xdhgz5fpe.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/615eb3cb7207f224b434de84

**Comment by ribel** [comment]
@navneet0693, @Ressinel All changes were made and the test pass. Can we merge it? :)

**Comment by Ressinel** [comment]
@rollki Could you update this PR from the main branch?

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2539-vab3fds8v2pqe6bjocasq64xdhgz5fpe.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/615eb3cb7207f224b434de84

**Comment by ribel** [comment]
@rollki @dmytriikaiun 
As we discussed with @navneet0693 we need one more change here.
In this PR we have `social_event_update_10202` and `social_event_update_10203`
But since we already have a stable 10.3.x releases with `social_event_update_10309`, those already updated projects will not receive this new event field.

So we agreed to change update numbers to `social_event_update_10310` and `social_event_update_10311`. 
And then we have two options on how to backport it to 10.2.x, since it already used as a patch:

1. Add a check if the field is already added in update hooks
2. Cherry pick to 102x but while cherry-picking change hook_update numbers to 102xx

Credits to @Ressinel who spotted it ;)

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2539-vab3fds8v2pqe6bjocasq64xdhgz5fpe.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/615eb3cb7207f224b434de84

**Comment by dmytriikaiun** [comment]
@navneet0693 update hook numbers were updated. Please check is all good and ready to merge.

File: modules/social_features/social_event/social_event.post_update.php [hunk 1]
```diff
@@ -54,3 +54,75 @@ function social_event_post_update_10301_enable_event_enrollment(&$sandbox) {
...
+
+/**
+ * Sets all day value for events.
+ *
+ * Now we do not use the time check if it is an event for all-day, we use the
+ * 'All day' field in event, in which case we must make an entry in database
+ * (State API) for all events that were for all-day, use the old logic for
+ * checking.
+ */
+function social_event_post_update_10204_set_all_day_value(array &$sandbox): void {
```
Comment (navneet0693) [inline modules/social_features/social_event/social_event.post_update.php:66]: Maybe change this one too 🙈.

`social_event_post_update_10302_set_all_day_value`

**Review by navneet0693 — CHANGES_REQUESTED** [review]

File: modules/social_features/social_event/social_event.post_update.php (hunk 1 above)
Comment (dmytriikaiun) [reply modules/social_features/social_event/social_event.post_update.php]: Done

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2539-vab3fds8v2pqe6bjocasq64xdhgz5fpe.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/615eb3cb7207f224b434de84

**Review by Ressinel — APPROVED** [review]
Looks good to me!

**Comment by navneet0693** [comment]
🍒 -picked to 10.3.x

**Pull request merged**